jniFaiss/src/com_amazon_opendistroforelasticsearch_knn_index_faiss_v164_KNNFIndex.cpp

```diff
@@ -0,0 +1,272 @@
+#include "com_amazon_opendistroforelasticsearch_knn_index_faiss_v164_KNNFIndex.h"
+
+#include <cmath>
+#include <cstdio>
+#include <cstdlib>
+#include <string>
+#include <vector>
+#include <sys/time.h>
+#include <omp.h>
+
+#include "faiss/index_factory.h"
+#include "faiss/MetaIndexes.h"
+#include "faiss/index_io.h"
+#include "faiss/IndexHNSW.h"
+
+
+using std::string;
+using std::vector;
+
+std::unordered_map<string, faiss::MetricType> mapMetric = {
+        {"l2", faiss::METRIC_L2},
+        {"innerproduct", faiss::METRIC_INNER_PRODUCT}
+};
+
+extern "C"
+
+struct JavaException {
```

Comment:
Can you add a TODO for future to abstract out common logic from JNI levels? This does not have to be completed in this PR, but we should do it in the future.